pulsar-io/core/src/main/java/org/apache/pulsar/io/core/SinkContext.java

```diff
@@ -40,4 +45,66 @@
      * @param value The value of the metric
      */
     void recordMetric(String metricName, double value);
+
+    /**
+     * Get a list of all input topics
+     * @return a list of all input topics
+     */
+    Collection<String> getInputTopics();
+
+    /**
+     * The tenant this sink belongs to
+     * @return the tenant this sink belongs to
+     */
+    String getTenant();
+
+    /**
+     * The namespace this sink belongs to
+     * @return the namespace this sink belongs to
+     */
+    String getNamespace();
+
+    /**
+     * The name of the sink that we are executing
+     * @return The Sink name
+     */
+    String getSinkName();
+
+    /**
+     * The logger object that can be used to log in a sink
+     * @return the logger object
+     */
+    Logger getLogger();
+    
+    /**
+     * Increment the builtin distributed counter refered by key
+     * @param key The name of the key
+     * @param amount The amount to be incremented
+     */
+    void incrCounter(String key, long amount);
```

Comment:
do we need these state methods for sink?

Reply:
are there any adverse effects of doing so? There could be use cases users what to use state in the sources and sinks.

Reply:
from implementation perspective, in current implementation, we construct a state context for each function call. so I don't think it will work in sink and source, since the context is in instance level. 

from semantic perspective, as you discussed in #3396, since and source context is instance level, while context is per function call. so it sounds to me they are different contexts. are we going to make them same as each other?

Reply:
I'm tending towards @sijie for this one. Not sure we really need this for sources/sinks

Reply:
@sijie the state context doesn't look like its constructed on a per message basis:

https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L240

Am I missing something?

Reply:
@srkukarni there could be many use cases for state for sources and sinks.  Like using state as a cache to store static data that needs to be loaded on start up as in functions.  I would also imagine that some connectors will need to be stateful in the future?

Reply:
It could well be useful in the future. However I always tend to add things only when needed. Thus unless there is a compelling usecase right now for sources/sinks to use this, my vote would be to defer it to later.

Reply:
@sijie I am not proposing we make Function context the same as Source/Sink Context because there are those differences as you mentioned.  However, in Function context, there are many instance level attributes that we can also expose in Source/Sink context.